### ISSUE TITLE
First try | Uboot won't find and load ubldr

### ISSUE THREAD

**Issue by yzgyyang** [yzgyyang/freebsd-noobs#1]
Following the [NOOBS partitioning rules](https://github.com/yzgyyang/freebsd-noobs/issues/), I created the following filesystems on an SD card, using the latest [Snapshot Build for RPI-B](https://download.freebsd.org/ftp/snapshots/ISO-IMAGES/12.0/):
 
RECOVERY (FAT, noobs)
- SETTINGS (EXT4, logical, noobs)
- boot (FAT, logical, uboot/ubldr) (directly copied from the official image)
- freebsd (BSD, logical, FreeBSD) (directly copied from the official image)

And then uboot is up without any configuration. However, it seems that it has no idea where itself is at, since it can't find ubldr in the same partition:

```
Booting from: mmc 0 ubldr.bin
reading ubldr.bin
** Unable to read file ubldr.bin **
U-Boot>
```

**Comment by emaste** [comment]
It also cannot find `uEnv.txt`. It looks like uboot is hard-coded to look at the first partition, which cannot work with NOOBS.

**Comment by yzgyyang** [comment]
> It looks like uboot is hard-coded to look at the first partition, which cannot work with NOOBS.

Did a simple test by throwing uboot stuff into the first primary partition (also FAT, which NOOBS uses as RECOVERY disk), then uboot found the loader.

**Comment by evadot** [comment]
Our u-boot-rpi ports is old and don't track u-boot-master, we will need to change that as the u-boot-master use the distroboot_cmd logic from u-boot that search every bootable source.
I the mean time I don't know how to configure u-boot to search in the right partition as, if I understood correctly, the partition can change based on how many OS NOOBS install.

**Comment by emaste** [comment]
> the partition can change based on how many OS NOOBS install

That is correct, NOOBS will create a FAT partition for the FreeBSD boot bits, but the partition number is arbitrary.

**Comment by yzgyyang** [comment]
Having a uEnv.txt in the main fat partition with fatdev=mmc 0:X (6 in this case) worked for uboot to have ubldr. Now I got:

```
FreeBSD/armv6 U-Boot loader, Revision 1.2
(Tue Jun 13 03:57:17 UTC 2017 root@releng3.nyi.freebsd.org)

DRAM: 480MB
Number of U-Boot devices: 2
U-Boot env: loaderdev='mmc 0:6'
Found U-Boot device: disk
  Checking unit=0 slice=6 partition=<auto>...
```

and it just stucks there.

**Comment by evadot** [comment]
As said on IRC I'm not sure that ubldr/loader(8) can work with logical partition (which are needed for partition number >4 in a MBR env)
A gpart show <device> from a FreeBSD host will help to see how things are done in NOOBS.

**Comment by yzgyyang** [comment]
```
➜  / gpart show

...

=>      63  60367809  da5  MBR  (29G)
        63      8129       - free -  (4.0M)
      8192    129024    1  !14  (63M)
    137216  60230656    2  ebr  (29G)

=>       0  60230656  da5s2  EBR  (29G)
         0     67582      1  linux-data  (32M)
     67582    135170   1073  !12  (66M)
    202752   2062376   3219  freebsd  (1.0G)
   2265128  57965528         - free -  (28G)

=>      0  2062313  da5s7  BSD  (1.0G)
        0      105         - free -  (53K)
      105  2062080      1  freebsd-ufs  (1.0G)
  2062185      128         - free -  (64K)
```

**Comment by evadot** [comment]
Ok so without looking at the code (but having read it many times) I'm pretty sure that loader(8) doesn't know about this partition scheme

**Comment by emaste** [comment]
From IRC:
```
21:35 <@DamnHippi> manu_, charlie it might be best to set loadderv=disk and let
                   ubldr do all the searching.  by setting it to 0:6 it's only 
                   looking there.
21:35 <@DamnHippi> ubldr knows how to search all mbr and ebr entries looking 
                   for an active freebsd partition, or just the first freebsd 
                   partition if none is active.
21:36 < manu_> DamnHippi: the problem is that NOOBS is using EBR partition
21:36 < manu_> DamnHippi: and I'm pretty sure that we cannot read that
21:36 <@DamnHippi> and it'll cope with the situation where the fat partition is 
                   marked active
21:36 <@DamnHippi> ubldr can handle ebr
21:36 < manu_> really ?
21:36 < manu_> ok
21:36 <@DamnHippi> right now it's failing because it was told to search exaclty 
                   one partition, and it wasn't a freebsd one.
21:38 < manu_> right
```

**Comment by emaste** [comment]
Another good point from IRC: the Windows boot must be using UEFI, so lets find out how they get into their UEFI loader.

**Comment by emaste** [comment]
Looking back at https://github.com/raspberrypi/noobs/issues/412,
> It would be nice to have a standard solution for U-BOOT that we can promote since many of the 64-bit OSes use it with a hard coded partition number (2) to load the rootfs from.

**Comment by bsdimp** [comment]
The first message has a link not to how to create an image, but to this file.

**Comment by emaste** [comment]
Probably https://github.com/raspberrypi/noobs/wiki/NOOBS-partitioning-explained

**Comment by yzgyyang** [comment]
> Another good point from IRC: the Windows boot must be using UEFI, so lets find out how they get into their UEFI loader.

I found this: https://github.com/ms-iot/RPi-UEFI

It seems to me that the onboard bootloader launches the UEFI environment and UEFI applications, then the UEFI environment launches the Windows Boot Manager.

**Comment by yzgyyang** [comment]
Giving `loaderdev=disk` and `fatdev=mmc 0:6` in uEnv.txt resulted in ubldr crashed at the same point:

```
FreeBSD/armv6 U-Boot loader, Revision 1.2
(Tue Jun 13 03:57:17 UTC 2017 root@releng3.nyi.freebsd.org)

DRAM: 480MB
Number of U-Boot devices: 2
U-Boot env: loaderdev='disk'
Found U-Boot device: disk
  Checking unit=0 slice=<auto> partition=<auto>...
```

I took the following log from a successful boot of the official image:
```
...
  Checking unit=0 slice=<auto> partition=<auto>... good.
Booting from disk0s2a:
/boot/kernel/kernel data=0x64cd68+0x1d7298 syms=[0x4+0xd7710+0x4+0x9ed9f]
...
```

**Comment by yzgyyang** [comment]
I am now looking at and trying to duplicate the way Microsoft does it using EDK2 UEFI. I am also interested in learning how our version of uboot and ubldr works. Do you guys have any insights? @emaste @evadot @bsdimp

**Comment by emaste** [comment]
For rpi3 see https://svnweb.freebsd.org/ports/head/sysutils/u-boot-rpi3/ for the u-boot port, and sys/boot/efi/ for our EFI boot bits.

For rpi2 https://svnweb.freebsd.org/ports/head/sysutils/u-boot-rpi2/ and sys/boot/arm/uboot/ for the ubldr bits.

**Comment by yzgyyang** [comment]
```
➜  gpart show            
...                                      
=>      63  30881729  da5  MBR  (15G)         
        63      8129       - free -  (4.0M)   
      8192    129024    1  !14  (63M)         
    137216  27326464    4  ebr  (13G)         
  27463680      4096       - free -  (2.0M)   
  27467776    131070    2  !12  (64M)         
  27598846         2       - free -  (1.0K)   
  27598848   3276800    3  ntfs  (1.6G)       
  30875648      6144       - free -  (3.0M)   

=>       0  27326464  da5s4  EBR  (13G)       
         0     67582      1  linux-data  (32M)                                              
     67582   1638400   1073  linux-data  (800M)                                             
   1705982      8192  27080  !112  (4.0M)     
   1714174  25608192  27210  ntfs  (12G)      
  27322366      4098  433689  !12  (2.0M)     
```
Ugh, meanwhile, Noobs-installed Windows IoT has 6 partitions which all seem to be involved in part of the booting process. So I installed Windows IoT without Noobs.

![screenshot at 2017-06-22 11 45 20](https://user-images.githubusercontent.com/16748524/27443212-575c8fdc-5740-11e7-8adf-3825791db375.png)

The brief introduction can be found here: https://www.blackhat.com/docs/us-16/materials/us-16-Sabanal-Into-The-Core-In-Depth-Exploration-Of-Windows-10-IoT-Core-wp.pdf

**Comment by yzgyyang** [comment]
Unfortunately, the Microsoft environment failed when I substitute bootmgfw.efi with FreeBSD'
s efi executable. There must be some kinds of checksum validation stuff going on. Also the serial connection doesn't work, regardless of which baud speed (I've tried a lot). Go back to uboot, hmm...

**Comment by procount** [comment]
Have you made any more progress on this? It seems to have gone quiet.

Maybe you already know, but after NOOBS installs an OS, it runs a partition_setup.sh script that you need to write. It's arguments indicate on which partitions (or partuuids) that the OS has been loaded onto. This is typically used to patch the cmdline.txt and fstab files so that the OS can be loaded from the correct partitions when it boots. I don't know UBoot well, but I believe there are many ways to indicate where the boot and root partitions are. Including such details in the UBOOT image is difficult to alter from the script, but configuring it to be external to the image would make life a lot easier.
